letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +78,108 @@ def validate_domain(self, domain):
-            'letsencrypt.reload_command', False)
-        if reload_cmd:
-            _logger.info('reloading webserver...')
-            self.call_cmdline(['sh', '-c', reload_cmd])
+        # Generate account key
+        account_key_file = self._generate_key('account_key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        net = client.ClientNetwork(account_key)
+        if self.env.context.get('DRY_RUN'):
+            directory_url = V2_STAGING_DIRECTORY_URL
+        else:
+            directory_url = V2_DIRECTORY_URL
+        directory_json = requests.get(directory_url).json()
+        acme = client.ClientV2(directory_json, net)
+        reg = Registration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        new_reg = NewRegistration(**dict(reg))
+        acme.new_account(new_reg)
+        altnames = parameter_model.search(
+            [('key', 'like', 'letsencrypt.altname.')],
+            order='key')
+        valid_domains = [x.value for x in altnames] + [domain]
+        _logger.info(
+            'Making CSR for the following domains: ' + str(valid_domains))
+        csr = crypto_util.make_csr(
+            open(domain_key_file).read(), valid_domains)
+        authzr = acme.new_order(csr)
+        # write the file with the challenge
+        challenge_dir = _get_challenge_dir()
+        if not isdir(challenge_dir):
+            makedirs(challenge_dir)
+        # if the domain contains a wildcard, do dns validation
+        if '*' not in domain:
+            challenge = filter(
+                lambda x: x.chall.typ == 'http-01',
+                authzr.authorizations[0].body.challenges)[0]
+            challenge_expiration = datetime.now() + timedelta(minutes=5)
+            token = base64.urlsafe_b64encode(challenge.token)
+            challenge_file = join(_get_challenge_dir(), '%s' % token)
+            with open(challenge_file, 'wb') as challenge_file:
+                # remove the trailing equals sign (=)
+                challenge_file.write(token + '.' + base64.urlsafe_b64encode(
+                    account_key.thumbprint())[:-1])
         else:
-            _logger.info('no command defined for reloading webserver, please '
-                         'do it manually in order to apply new certificate')
+            challenge_expiration = authzr.authorizations[0].body.expires
+            token = authzr.authorizations[0].body.challenges[0].token
+            # both challenges are DNS challenges, pick the first one
+            token = base64.b64encode(token)
+            # create the dns_record
+            ir_config_parameter = self.env['ir.config_parameter']
+            letsencrypt_dns_function = 'letsencrypt_' + \
+                ir_config_parameter.get_param('letsencrypt_dns_provider') + \
+                '_support'
```

Comment:
here I think you should just query the provider, and call `self._letsencrypt_dns_auth_$provider` which then does whatever is necessary

Reply:
Can you help me out on this one? I do not seem to understand how do you want me to refactor this.

Reply:
I have something like this in mind (pseudo code)

```python
dns_provider = self.env['ir.config_parameter'].get_param(...)
dns_update_function = getattr(self, '_letsencrypt_dns_auth_%s' % dns_provider)
# this function then in your case would read the two parameters and call the api function
dns_update_function(challenge_expiration, token)
# or just pass the authz, you know the acme protocol better than me to judge what other code might need
# dns_update_function(authzr)
```

so the idea is: Pass all relevant information you got from letsencrypt to the function implementing the dns provider, and expect it to do the right thing. Also document this interface in the readme.